napi-inl.h

```diff
@@ -4112,6 +4114,156 @@ inline void ThreadSafeFunction::CallJS(napi_env env,
+                                const Function& callback)
+  : AsyncProgressWorker(receiver, callback, "generic") {
+}
+
+template<class T>
+inline AsyncProgressWorker<T>::AsyncProgressWorker(const Object& receiver,
+                                const Function& callback,
+                                const char* resource_name)
+  : AsyncProgressWorker(receiver,
+                callback,
+                resource_name,
+                Object::New(callback.Env())) {
+}
+
+template<class T>
+inline AsyncProgressWorker<T>::AsyncProgressWorker(const Object& receiver,
+                                                   const Function& callback,
+                                                   const char* resource_name,
+                                                   const Object& resource)
+  : AsyncWorker(receiver, callback, resource_name, resource),
+    _asyncdata(nullptr),
+    _asyncsize(0) {
+  _tsfn = ThreadSafeFunction::New(callback.Env(), callback, resource_name, 1, 1);
+}
+
+#if NAPI_VERSION > 4
+template<class T>
+inline AsyncProgressWorker<T>::AsyncProgressWorker(Napi::Env env)
+  : AsyncProgressWorker(env, "generic") {
+}
+
+template<class T>
+inline AsyncProgressWorker<T>::AsyncProgressWorker(Napi::Env env,
+                                const char* resource_name)
+  : AsyncProgressWorker(env, resource_name, Object::New(env)) {
+}
+
+template<class T>
+inline AsyncProgressWorker<T>::AsyncProgressWorker(Napi::Env env,
+                                const char* resource_name,
+                                const Object& resource)
+  : AsyncProgressWorker(env, resource_name, resource),
+    _asyncdata(nullptr),
+    _asyncsize(0) {
+  // FIXME: Napi::ThreadSafeFunction nullptr callback support?
+  Function callback;
+  _tsfn = ThreadSafeFunction::New(env, callback, resource_name, 1, 1);
+}
+#endif
+
+template<class T>
+inline AsyncProgressWorker<T>::~AsyncProgressWorker() {
+  // Abort pending tsfn call.
+  // Don't send progress events after we've already completed.
+  _tsfn.Abort();
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _asyncdata = nullptr;
+    _asyncsize = 0;
+  }
+  _tsfn.Release();
+}
+
+template<class T>
+inline void AsyncProgressWorker<T>::Execute() {
+  ExecutionProgress progress(this);
+  Execute(progress);
+}
+
+template<class T>
+inline void AsyncProgressWorker<T>::WorkProgress_(Napi::Env /* env */, Napi::Function /* jsCallback */, void* _data) {
+  AsyncProgressWorker* self = static_cast<AsyncProgressWorker*>(_data);
+
+  T* data;
+  size_t size;
+  {
+    std::lock_guard<std::mutex> lock(self->_mutex);
+    data = self->_asyncdata;
+    size = self->_asyncsize;
+    self->_asyncdata = nullptr;
+    self->_asyncsize = 0;
+  }
+
+  self->OnProgress(data, size);
+  delete[] data;
+}
+
+template<class T>
+inline void AsyncProgressWorker<T>::SendProgress_(const T* data, size_t count) {
+    T* new_data = new T[count];
+    {
+      T* it = new_data;
```

Comment:
Why do we need a copy of the pointer here versus just using new_data in the copy?

Looks like this does not add to existing data, just swaps in the new data and deletes the old data. From the doc earlier I might have thought it needed to add it to the existing data.

Reply:
There are a few cases of how the data pointer were managed:

1. the data was created on the stack, that is how we tested it. In this case, the data has to be copied to be used on the main thread.
2. the data was created on the heap, there is no test on that but since there are no finalizers to free the data pointer until the main thread usage is done, it would be safer to just copy the data and have it managed by the Worker itself. 

With acknowledged above two conditions, I'd keep the API consistent with the NAN one.

Reply:
Sorry I think my question was not clear. I understand why we need to copy the data.  The question is why we need to use the T* it pointer instead of just using new_data in the copy instead of `it`?

Reply:
Ah, yes, I think it is unnecessary too. Just copied from NAN implementation. Submitted a PR to NAN: https://github.com/nodejs/nan/pull/870.